File: _a_arm/python_old/hw8/armController.py
import numpy as np
import armParamHW8 as P
import sys
sys.path.append('..')  # add parent directory
import armParam as P0
from PDControl import PDControl


class armController:

    def __init__(self):
        # Instantiates the PD object
        self.thetaCtrl = PDControl(P.kp, P.kd, P0.tau_max, P.beta, P.Ts)
        self.limit = P0.tau_max

    def u(self, y_r, y):
        # y_r is the referenced input
        # y is the current state
        theta_r = y_r[0]
        theta = y[0]

        # compute equilibrium torque tau_e
        tau_e = P0.m * P0.g * (P0.ell / 2.0) * np.cos(theta)
        # compute the linearized torque using PD
        tau_tilde = self.thetaCtrl.PD(theta_r, theta, False)
        # compute total torque
        tau = tau_e + tau_tilde
        tau = self.saturate(tau)
        return [tau]

    def saturate(self, u):
        if abs(u) > self.limit:
            u = self.limit*np.sign(u)
        return u

File: _a_arm/matlab_old/hw7/armController.m
classdef armController
    % 
    %    This class inherits other controllers in order to organize multiple controllers.
    %
    %----------------------------
    properties
        thetaCtrl
        m
        ell
        g
    end
    %----------------------------
    methods
        %----------------------------
        function self = armController(P)
            % Instantiates the SS_ctrl object
            self.thetaCtrl = PDControl(P.kp, P.kd, P.tau_max, P.beta, P.Ts);
            % plant parameters known to controller
            self.m = P.m;
            self.ell = P.ell;
            self.g = P.g;
        end
        %----------------------------
        function tau = u(self, y_r, y)
            % y_r is the referenced input
            % y is the current state
            theta_r = y_r;
            theta = y(1);
            
            % compute equilibrium torque tau_e
            tau_e = self.m*self.g*(self.ell/2)*cos(theta);
            % compute the linearized torque using PID
            tau_tilde = self.thetaCtrl.PD(theta_r, theta, false);
            % compute total torque
            tau = tau_e + tau_tilde;
        end
    end
end